config/parser.go

```diff
@@ -0,0 +1,89 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+type parser struct {
+	dups map[Source]string
+}
+
+func newParser() *parser {
+	return &parser{}
+}
+
+func (p *parser) GetFields(cfg interface{}) ([]*Field, error) {
+	p.dups = make(map[Source]string)
+
+	tp := reflect.TypeOf(cfg)
+	if tp.Kind() != reflect.Ptr {
+		return nil, errors.New("configuration should be a pointer type")
+	}
+	val := reflect.ValueOf(cfg).Elem()
+
+	return p.getFields("", tp.Elem(), &val)
+}
+
+func (p *parser) getFields(prefix string, tp reflect.Type, val *reflect.Value) ([]*Field, error) {
+	var ff []*Field
+	for i := 0; i < tp.NumField(); i++ {
+		f := tp.Field(i)
+		fld, err := NewField(prefix, &f, val)
```

Comment:
I believe it would be better for readability if we:
- Check if the field is nested, if yes then call getFields with the appropriate prefix
- If not just create a `NewField`
this way the code is more readable and we could support unlimited nesting (recursion).
We should also provide a test with 2 level deep testing for this.

Reply:
Well, there is a design idea behind this implementation. For now nesting is a fallback approach when parser is unable to create a Field. Usually it could be because of two reasons (both are being checked inside `NewField` func): 
1. Struct field type is not a struct 
OR
2. Underlying structure type is not from `sync` package 

And this implementation details are logically incapsulated close to `Field`, so we just use "tell don't ask" approach and on parser side we have no idea from what kind of struct fields our `Field` structs could be created. And only in case of failure we check for nesting.

In general this check "is this field is nested" means -- Is this struct field can not be a`Field` AND this struct field is a struct itself. And an easiest way to figure it out - just try to convert it to `Field`. But perhaps I'm missing something important, not sure.

Ah, forgot to mention - current implementation already supports unlimited nesting levels. I'll add a test for this.

Reply:
TBH returning and error and make code branch decision to parse structures, seems like a bad idea.
The error is not even a specific struct error in order to check for that and then make the branch.
If would propose to have a simple if check to see if it is a struct or not and handle the field accordingly.

I have another idea. Would it make sense to not allow the nested struct to be a pointer and return an error? Since this is a config struct there is possibly no reason for having a pointer in the first place. what do you think?

I did totally overlooked the unlimited nesting part..

Reply:
You're totally right, the idea of using non-specific errors to make logical decisions is not explicit enough. To make this part more linear and readable (and a lot shorter) and using your idea of limiting to use only structs without pointers we can do something like this (in pseudo go):
```go
if fieldType.Kind() != reflect.Struct {
  return onlyStructsAllowedError
}
// Since all we can have in config is structs, 
// we still need to determine what do we have - 
// potential Field or just nested struct
if isFieldTypeSupported(fieldType) {
   newField := createField(...)
} else {
   nestedFields := getFields(...)
}
```
And in this case we will have simple and explicit logic. What do you thing about this approach?

Reply:
We can avoid the error totally if we di the following after line 32:
- Check if `f` is valid. I would introduce a validation function to do this. There I would check if the field had our tags it should be using our types from the sync package and pointers to struct. The function would return an error and the type.
- Using the type I would check if it is a struct and call the appropriate function else to call the normal field creation.
This approach, I believe:
- makes the code much more readable because everything is explicit.
- does not need special error types to be handled
What do you think?

Reply:
Yes, it feels like better approach. Just committed new version of parser implementation. 

Not sure if it is now implemented exactly as you've proposed and it feels a little complex because of additional types (field/struct), but in general it looks much more readable because of explicit checks and linear logic. 

Could you please take a look? Thanks a lot for your impact!

Reply:
I believe this looks really nice and is much more readable. 
Very nicely done. Kudos.